### DISCUSSION TITLE
Improve caching of assets by providing ETag/Last-Modified headers

### DISCUSSION THREAD

**Discussion by DanielBiegler** [directus/directus#19768] [Feature Requests]
### Summary

This would improve response times by being able to reply with 304 status codes in case the resource hasnt changed. This then allows us to lower the default `max-age` value so we can discover changed assets quicker **without** paying for the request re-downloading the full asset.

### Basic Example

Browsers will send us `if-modified-since` and `if-none-matched` headers that we can get like so:

```js
const ifModifiedSince = req.headers?.['if-modified-since'];
```

and after constructing some Date objects we can check for them

```js
if(lastModifiedDate > askedModifiedDate === false) {
	return res.status(304).end();
}
```

### Motivation

See summary.

### Detailed Design

1. Currently we actually _do_ remove all generated transformations when replacing a file (v10.6.2)
2. We do serve the new file as per my screenshot [here](https://github.com/directus/directus/issues/17477#issuecomment-1731096561).

Technically, if

- you provide cache disabling headers for `/assets`
- the browser ignores caching itself
- you use your own cachebusting param with a new value, for example `/assets/<hash>?cachebust=abc123`

the browser **will** see the new resource. Going further along without the mentioned options:

3. The new transformations get persisted once upon request for the new resource
4. Now the browser refuses to replace the cached resource because it's still [fresh](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#fresh_and_stale_based_on_age) defined by us setting the [max-age](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#response_directives).
5. max-age seems to have a higher priority than other caching options 

### What we could do

1. Get rid off the max-age header completely and switch to a [conditional request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests) where we can reply with an [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) and or Last-Modified header to either return the resource or reply with body-less and tiny [304 Not Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304)

This'd allow to always get the latest resource but with the cost of an extra request. Probably for most people the following would be best:

2. Use either the ETag and or modified-headers to be able to reply with tiny 304 responses, but make the max-age default from [ASSETS_CACHE_TTL](https://docs.directus.io/self-hosted/config-options.html#assets) **way shorter** than the current default of 30days. This'd allow for good prevention of unnecessary requests and the quick revalidation via the 304 reply where we don't have to transmit the image at all after max-age.

Messing with caches can and will impact every Directus instance so I'd love to be very careful with this.

### Requirements List

See detailed design.

### Drawbacks

None worth mentioning.

### Alternatives

- You can just cachebust yourself via a parameter
- You can decrease [ASSETS_CACHE_TTL](https://docs.directus.io/self-hosted/config-options.html#assets) and depending on your traffic it doesnt matter.

### Adoption Strategy

It's not a breaking change and a simple update.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
